GRDB: ensure FMDB and GRDB compatibility

leandroalonso opened the pull request:
| 📘 Part of: #2773 |
|:---:|

While doing the process to migrate from FMDB to GRDB we want something simple: that we can alternate between FMDB and GRDB as needed and that this doesn't cause any issue.

#### Issue 1: Dates

While working on #2820, I discovered that FMDB and GRDB handle dates differently. FMDB saves it using `timeIntervalSince1970` while GRDB uses "the format "YYYY-MM-DD HH:MM:SS.SSS" in the UTC time zone. It is precise to the millisecond." (from their docs).

GRDB is much better, however — at this stage, we want compatibility between them. So this PR forces GRDB to also save using `timeIntervalSince1970`.

#### Issue 2: usage of double quotes

There was a query using double quotes. This should not be done and causes an error that was caugh by a unit test, you can see the [change here](https://github.com/Automattic/pocket-casts-ios/pull/2840/commits/227b0d83f1615af9160c8d7327357091ad68c502).

More context about [this issue here](https://github.com/Automattic/pocket-casts-ios/issues/2773#issuecomment-2721222352).

#### More issues?

We might have more issues. The only way to know is by using GRDB. So my proposal is to enable it first for our development environment (using `#if DEBUG`).

## Changes

* Extract `Episode` and `Podcast` creation from a dictionary to its own method (so it can be used in unit tests)
* Update the old code where the extracted code lived to call the new methods
* Change `GRDBDatabase.executeUpdate` to parse the given values and change `Date` to `TimeInterval`
* Add unit tests that compare that the same podcast/episode saved from GRDB and FMDB produces the same object
* Add unit tests to ensure that podcasts/episodes saved with GRDB and read from FMDB produces the same results
* Add unit tests to ensure that podcasts/episodes saved with FMDB and read from GRDB procudes the same results
* Changed all unit tests to run with the enabled framework. If GRDB is disabled it uses FMDB, otherwise it uses GRDB
* Changed a SQL Query that was using double quotes and failing to execute when running using GRDB

## Discussion

I think adding unit tests for `Podcast` and `Episode` are enough because it coverts all the data types used by other entities (such as Bookmark, Folder, etc).

## To test

### From FMDB to GRDB

1. Run `trunk`
2. Checkout this branch
3. Run the app again
4. ✅ Check the console, it should have no database warnings
5. Check the podcasts and episodes, they should display correctly

### From GRDB to FMDB

1. Delete the app
2. Run this branch
3. Log in to an account with items in Up Next, followed podcasts, etc.
4. Change the `grdb` feature flag to `false`
5. Re-run the app
6. ✅ Check the console, it should have no database warnings
7. ✅ Check the podcasts, episodes, Up Next, it all should behave correctly

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

leandroalonso committed:
Extract episode creation

So it can be used in unit tests

leandroalonso committed:
Change to use Episode.from

leandroalonso committed:
Extract podcast creation from dictionary to a method

So it can be used in unit tests

leandroalonso committed:
Add a new method to create a database using GRDB

leandroalonso committed:
Add as an extension of DatabasePool

leandroalonso committed:
Create a podcast fixture

leandroalonso committed:
Create podcast and episodes in FMDB and GRDB and compare them

leandroalonso committed:
Compare podcasts read from each other DB

leandroalonso committed:
Remove unused var

leandroalonso committed:
Add episode tests

leandroalonso committed:
Ensure Date on Podcast/Episode are handled correct

Meaning: it's saved using timeIntervalSince1970

leandroalonso committed:
Revert "Ensure Date on Podcast/Episode are handled correct"

This reverts commit 6dff6cd6741542406cede25ec8c13c9ad2ef33cc.

leandroalonso committed:
Transform Date into TimeInterval

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

leandroalonso committed:
Do not use double quotes

This causes issues when using GRDB, as FMDB handles this for us.

leandroalonso committed:
Update all tests to run using the enabled framework

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility

leandroalonso committed:
Enable GRDB flag so CI runs tests with it

leandroalonso commented on Modules/Utils/Sources/PocketCastsUtils/Feature Flags/FeatureFlag.swift:
```diff
@@ -258,7 +258,7 @@ public enum FeatureFlag: String, CaseIterable {
             false
         case .grdb:
-            false
+            true
```
**This should not be merged with the flag enabled.**

My proposal is that we change this to:

```suggestion
            #if DEBUG
            true
            #else
            false
            #endif
```

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility

SergioEstevao commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/DB/GRDB/GRDBDatabase.swift:
```diff
@@ -17,7 +18,12 @@ class GRDBDatabase: PCDatabase {
+        // However, that's a large change that we want to avoid.
+        let filteredValues = values?.map { ($0 as? Date)?.timeIntervalSince1970 ?? $0 }
+
+        try database.execute(sql: sql, arguments: StatementArguments(filteredValues != nil ? filteredValues! : [])!)
```
Do we want to keep the force unwrapping for StatementArguments? 
Is this for us to detect any arguments that is not `DatabaseValueConvertible` ?

leandroalonso replied:
I would keep this way for now.

You can see a good example described in this comment: https://github.com/Automattic/pocket-casts-ios/pull/2840#issuecomment-2787429208

tl;dr: forcing unwrapping for GRDB made me discover underlying issues on our unit tests.

SergioEstevao commented:
When running the unit tests locally with the `grdb` FF active I'm seeing the following exceptions:

`
PocketCastsDataModel/GRDBQueue.swift:13: Fatal error: 'try!' expression unexpectedly raised an error: SQLite error 21: out of memory - while executing `BEGIN IMMEDIATE TRANSACTION`
`

Expected?

leandroalonso committed:
Fix wrong database usage based on flag

leandroalonso committed:
Clear download queue after tests

This was leaking the added episodes while other tests were running. With FMDB that was hidden, with GRDB - given we throw errors - that was clear

leandroalonso committed:
Fix an issue when creating a DataManager with GRDB

leandroalonso commented:
@SergioEstevao definetely not.

I've checked what was going on and there were multiple issues:

1. I've inverted the correct `DataManager` initialization. That is fixed in 7a0208c79328a1a686a4957c8b346eedf44b8ddd
2. By running the tests using GRDB — and forcing unwrapping — I discovered an issue with the download tests. Those were leaking while other tests were running, this is fixed in 55d33cb089a1e46b6e8067116638b5443568914c
3. The `DataManager` initialization for GRDB was wrong when the `grdb` flag was disabled. This was causing the database connection to be closed due to the Feature Flag being `false`. Fixed on 7a0208c79328a1a686a4957c8b346eedf44b8ddd

Tests should run correctly now, with `grdb` flag on/off.

SergioEstevao commented:
Tests are now running correctly!
However when running in GRDB mode, I got an exception when playing around with bookmarks.
After I dig at it a bit I found the exception was resulting from this line:
`
let result = await dbQueue.executeUpdate(query, values: [SyncStatus.synced])
`
on `BookmarkDataManager`

From what I could found the issue is that we are not calling rawValue on the enum when sending to the database and then the `StatementArguments` does not know how to convert it properly and return nil.

Changing it to this:
`
let result = await dbQueue.executeUpdate(query, values: [SyncStatus.synced.rawValue])
`
solves the issue, but we will need to keep an eye on the database call other places were we are using enums directly instead of using rawValue

leandroalonso commented:
@SergioEstevao That's the good thing about force unwraps while testing. :)

We should make `SyncStatus` to conform to `DatabaseConvertible`, I added this on #2840.

This PR is already quite big, and we will still find a few things here and here.

SergioEstevao reviewed (approved):
Now that unit tests are working correctly let's merge this one, and handle the other issues we found on follow up PRs.
:shipit:

danielebogo reviewed (approved):
🎉 🎉 🎉 🎉 🎉

leandroalonso committed:
Merge branch 'grdb/add-grdb-abstraction' into grdb/ensure-fmdb-grdb-compatibility
